caravel/assets/javascripts/SqlLab/actions.js

```diff
@@ -25,11 +25,119 @@ export const ADD_ALERT = 'ADD_ALERT';
 export const REMOVE_ALERT = 'REMOVE_ALERT';
 export const REFRESH_QUERIES = 'REFRESH_QUERIES';
 export const SET_NETWORK_STATUS = 'SET_NETWORK_STATUS';
+export const RUN_QUERY = 'RUN_QUERY';
+export const START_QUERY = 'START_QUERY';
+export const STOP_QUERY = 'STOP_QUERY';
+export const REQUEST_QUERY_RESULTS = 'REQUEST_QUERY_RESULTS';
+export const QUERY_SUCCESS = 'QUERY_SUCCESS';
+export const QUERY_FAILED = 'QUERY_FAILED';
+export const CLEAR_QUERY_RESULTS = 'CLEAR_QUERY_RESULTS';
+export const HIDE_DATA_PREVIEW = 'HIDE_DATA_PREVIEW';
 
 export function resetState() {
   return { type: RESET_STATE };
 }
 
+export function startQuery(query) {
+  Object.assign(query, {
+    id: shortid.generate(),
+    progress: 0,
+    startDttm: now(),
+    state: (query.runAsync) ? 'pending' : 'running',
+  });
+  return { type: START_QUERY, query };
+}
+
+export function querySuccess(query, results) {
+  return { type: QUERY_SUCCESS, query, results };
+}
+
+export function queryFailed(query, msg) {
+  return { type: QUERY_FAILED, query, msg };
+}
+
+export function stopQuery(query) {
+  return { type: STOP_QUERY, query };
+}
+
+export function clearQueryResults(query) {
+  return { type: CLEAR_QUERY_RESULTS, query };
+}
+
+export function hideDataPreview() {
+  return { type: HIDE_DATA_PREVIEW };
+}
+
+export function requestQueryResults(query) {
+  return { type: REQUEST_QUERY_RESULTS, query };
+}
+
+export function fetchQueryResults(query) {
+  return function (dispatch) {
+    dispatch(requestQueryResults(query));
+    const sqlJsonUrl = `/caravel/results/${query.resultsKey}/`;
+    $.ajax({
+      type: 'GET',
+      dataType: 'json',
+      url: sqlJsonUrl,
+      success(results) {
+        // console.log(results);
+        dispatch(querySuccess(query, results));
+      },
+      error() {
+        // console.log(err);
+        dispatch(queryFailed(query, 'Failed at retrieving results from the results backend'));
+      },
+    });
+  };
+}
+
+export function runQuery(query) {
+  return function (dispatch) {
+    dispatch(startQuery(query));
+    const sqlJsonUrl = '/caravel/sql_json/';
+    const sqlJsonRequest = {
+      client_id: query.id,
+      database_id: query.dbId,
+      json: true,
+      runAsync: query.runAsync,
+      schema: query.schema,
+      sql: query.sql,
+      sql_editor_id: query.sqlEditorId,
+      tab: query.tab,
+      // tmp_table_name: this.state.ctas,
+      select_as_cta: query.ctas,
+    };
+    $.ajax({
+      type: 'POST',
+      dataType: 'json',
+      url: sqlJsonUrl,
+      data: sqlJsonRequest,
+      success(results) {
+        if (!query.runAsync) {
+          dispatch(querySuccess(query, results));
+        }
+      },
+      error(err, textStatus, errorThrown) {
+        let msg;
```

Comment:
yeah I had the same feeling that we should generalize that somehow, we're supposed to start using `fetch` instead of jquery, maybe then would be a good time